src/LineBlockTracer.cpp

```diff
@@ -49,6 +50,160 @@ bool cLineBlockTracer::Trace(cWorld & a_World, cBlockTracer::cCallbacks & a_Call
 
 
 
+bool cLineBlockTracer::LineOfSightTrace(cWorld & a_World, const Vector3d & a_Start, const Vector3d & a_End, eLineOfSight a_Sight)
+{
+	switch (a_Sight)
+	{
+		case losAir:
+		{
+			static class LineOfSightCallbacksAir:
+				public cLineBlockTracer::cCallbacks
+			{
+			public:
+				LineOfSightCallbacksAir() {}
+
+				virtual bool OnNextBlock(int a_BlockX, int a_BlockY, int a_BlockZ, BLOCKTYPE a_BlockType, NIBBLETYPE a_BlockMeta, eBlockFace a_EntryFace) override
+				{
+					if (a_BlockType != E_BLOCK_AIR)
+					{
+						return true;  // Abort tracing
+					}
+					return false;
+				}
+			} callbacksAir;
+			return Trace(a_World, callbacksAir, a_Start, a_End);
+		}
+
+		case losAirWater:
+		{
+			static class LineOfSightCallbacksAirWater:
+				public cLineBlockTracer::cCallbacks
+			{
+			public:
+				LineOfSightCallbacksAirWater() {}
+
+				virtual bool OnNextBlock(int a_BlockX, int a_BlockY, int a_BlockZ, BLOCKTYPE a_BlockType, NIBBLETYPE a_BlockMeta, eBlockFace a_EntryFace) override
```

Comment:
It feels like the virtual function call is going to be a bigger part of that. I'm now wondering if there is any way we can get rid of that, would it be worth templating the tracer based on what blocks we are interested/not interested in?

Also the callbacks are massively ugly. I'd love to suggest changing the callbacks to use closures but I'm not sure thats possible without introducing another allocation.